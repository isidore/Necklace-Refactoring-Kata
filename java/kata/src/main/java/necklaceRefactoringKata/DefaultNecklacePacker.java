package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;

class DefaultNecklacePacker implements PackerApplesauce {
    @Override
    public boolean pack(Jewellery i, JewelleryStorage s) {
        return s.tree.add(i);
    }
}
