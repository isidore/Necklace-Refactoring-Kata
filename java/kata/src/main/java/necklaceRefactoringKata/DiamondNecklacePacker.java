package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;

public class DiamondNecklacePacker implements PackLogic {
    @Override
    public boolean pack(Jewellery item, JewelleryStorage storage) {
        if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
            return true;
        }
        return false;
    }
}
