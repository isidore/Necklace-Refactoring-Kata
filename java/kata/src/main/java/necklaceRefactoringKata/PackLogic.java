package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;

public interface PackLogic {
    boolean pack(Jewellery item, JewelleryStorage storage);
}
