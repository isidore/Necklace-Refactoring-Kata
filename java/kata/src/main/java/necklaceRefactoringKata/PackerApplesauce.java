package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;

public interface PackerApplesauce {
    boolean pack(Jewellery item, JewelleryStorage storage);
}
