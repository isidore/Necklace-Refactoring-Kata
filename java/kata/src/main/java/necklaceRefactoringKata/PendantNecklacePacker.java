package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.PendantNecklace;

public class PendantNecklacePacker implements PackLogic {
    @Override
    public boolean pack(Jewellery item, JewelleryStorage storage) {
        if (!(item instanceof PendantNecklace pendantNecklace)) {
            return false;
        }
        storage.tree.add(pendantNecklace.chain);
        storage.box.topShelf.add(pendantNecklace.pendant);
        return true;
    }
}
