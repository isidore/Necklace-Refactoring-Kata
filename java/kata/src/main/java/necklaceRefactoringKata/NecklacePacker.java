package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.packers.*;
import org.lambda.query.Queryable;

public class NecklacePacker implements PackLogic {
    @Override
    public boolean pack(Jewellery item, JewelleryStorage storage) {
        return Queryable.as(new DiamondNecklacePacker(),
                new SmallNecklacePacker(),
                new PendantNecklacePacker(),
                new DefaultNecklacePacker()).any(p -> p.pack(item, storage));
    }
}
