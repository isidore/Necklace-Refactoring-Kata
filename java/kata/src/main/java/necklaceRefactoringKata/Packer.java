package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Earring;
import necklaceRefactoringKata.jewellery.EarringType;
import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Necklace;
import necklaceRefactoringKata.jewellery.PendantNecklace;
import necklaceRefactoringKata.packers.*;

public class Packer {

    public static void packNecklace(Necklace item, JewelleryStorage storage) {
        packItem(item, storage, new DiamondNecklacePacker(),
                new SmallNecklacePacker(),
                new PendantNecklacePacker(),
                new DefaultNecklacePacker());
    }

    private static void packItem(Jewellery item, JewelleryStorage storage, PackLogic...packers) {
        for (PackLogic packer : packers) {
            if (packer.pack(item, storage)) {
                return;
            }
        }
    }

    public static boolean pack(Jewellery item, JewelleryStorage storage) {
        if (storage.IsInTravelRoll(item) && !item.isLarge()) {
            storage.box.topShelf.add(item);
            storage.travelRoll.remove(item);
            return true;
        } else if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
            return removeItem(item, storage);
        } else if (item.isSmall()) {
            storage.box.topShelf.add(item);
            return removeItem(item, storage);
        } else if (item instanceof Earring earring && earring.type == EarringType.Hoop) {
            storage.tree.add(earring);
            return removeItem(item, storage);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop && earring.stone != Jewel.Plain) {
            storage.box.topShelf.add(earring);
            return removeItem(item, storage);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop) {
            storage.box.mainSection.add(earring);
            return removeItem(item, storage);
        } else if (item instanceof PendantNecklace pendantNecklace) {
            storage.tree.add(pendantNecklace.chain);
            storage.box.topShelf.add(pendantNecklace.pendant);
            return removeItem(item, storage);
        } else if (item instanceof Necklace necklace) {
            storage.tree.add(necklace);
            return removeItem(item, storage);
        }

        packItem(item, storage, new DefaultJewelleryPacker());
        return true;
    }

    private static boolean removeItem(Jewellery item, JewelleryStorage storage) {
        if (storage.IsInTravelRoll(item)) {
            storage.travelRoll.remove(item);
        }
        return true;
    }

}
