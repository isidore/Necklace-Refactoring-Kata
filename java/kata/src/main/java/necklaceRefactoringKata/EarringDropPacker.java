package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Earring;
import necklaceRefactoringKata.jewellery.EarringType;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.packers.PackLogic;

public class EarringDropPacker implements PackLogic {
    @Override
    public boolean pack(Jewellery item, JewelleryStorage storage) {
        if (!(item instanceof Earring earring) || earring.type != EarringType.Drop) {
            return false;
        }
        storage.box.mainSection.add(earring);
        return true;
    }
}
