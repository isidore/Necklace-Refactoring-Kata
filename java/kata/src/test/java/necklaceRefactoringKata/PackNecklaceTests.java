package necklaceRefactoringKata;


import necklaceRefactoringKata.jewellery.*;
import org.approvaltests.Approvals;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Test;

class PackNecklaceTests {
    @Test
    void testPackNecklace() {
        var item = new Necklace(Jewel.Pearl, NecklaceType.Beads);
        var storage = new JewelleryStorage();

        Packer.packNecklace(item, storage);
        Packer.packNecklace(new Necklace(Jewel.Diamond, NecklaceType.Beads), storage);
        Packer.packNecklace(new Necklace(Jewel.Amber, NecklaceType.Chain), storage);
        Jewel jewel = Jewel.Plain;
        NecklaceType necklaceType = NecklaceType.LongChain;
        Necklace necklace = new Necklace(jewel, necklaceType);
        Jewellery jewellery = new Pendant(Jewel.Pearl);
        Packer.packNecklace(new PendantNecklace(jewel, necklaceType, necklace, jewellery), storage);

        JsonApprovals.verifyAsJson(storage);
        // TODO: check if packed in right container
    }
    @Test
    void testPack() {
        var storage = new JewelleryStorage();

        var jewelleries = new Jewellery[]{
                new Necklace(Jewel.Pearl, NecklaceType.Beads),
                new Necklace(Jewel.Diamond, NecklaceType.Beads),
                new Necklace(Jewel.Amber, NecklaceType.Chain),
                new PendantNecklace(Jewel.Plain,
                        NecklaceType.LongChain,
                        new Necklace(Jewel.Plain, NecklaceType.LongChain),
                        new Pendant(Jewel.Pearl)),
                new Earring(Jewel.Amber, EarringType.Stud),
                new Earring(Jewel.Amber, EarringType.Hoop),
                new Earring(Jewel.Amber, EarringType.Drop),
                new Earring(Jewel.Plain, EarringType.Drop),
        };
        for (Jewellery j : jewelleries) {
            Packer.pack(j, storage);
        }

        JsonApprovals.verifyAsJson(storage);
    }
}
