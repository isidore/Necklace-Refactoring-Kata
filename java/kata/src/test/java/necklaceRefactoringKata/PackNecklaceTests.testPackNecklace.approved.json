{
  "box": {
    "ringCompartment": [],
    "topShelf": [
      {
        "type": "Chain",
        "stone": "Amber"
      },
      {
        "stone": "Pearl"
      }
    ],
    "mainSection": []
  },
  "tree": [
    {
      "type": "Beads",
      "stone": "Pearl"
    },
    {
      "type": "LongChain",
      "stone": "Plain"
    }
  ],
  "travelRoll": [],
  "safe": [
    {
      "type": "Beads",
      "stone": "Diamond"
    }
  ],
  "dresserTop": []
}